File: macos/Tests/YOLOTermKitTests/LayoutEngineTests.swift
import XCTest
@testable import YOLOTermKit

final class LayoutEngineTests: XCTestCase {
    
    let engine = LayoutEngine()
    
    // MARK: - Fixture Tests
    
    func testSinglePane() throws {
        // Load fixture: contracts/fixtures/layout/single-pane.json
        let fixture = try loadLayoutFixture(named: "single-pane")
        
        let result = engine.calculate(
            paneIds: fixture.input.panes.map { $0.id },
            preset: fixture.input.preset,
            containerSize: fixture.input.containerSize
        )
        
        assertLayoutMatches(result: result, expected: fixture.expected)
    }
    
    func testTwoColumns() throws {
        // Load fixture: contracts/fixtures/layout/two-columns.json
        let fixture = try loadLayoutFixture(named: "two-columns")
        
        let result = engine.calculate(
            paneIds: fixture.input.panes.map { $0.id },
            preset: fixture.input.preset,
            containerSize: fixture.input.containerSize
        )
        
        assertLayoutMatches(result: result, expected: fixture.expected)
    }
    
    func testFourGrid() throws {
        // Load fixture: contracts/fixtures/layout/four-grid.json
        let fixture = try loadLayoutFixture(named: "four-grid")
        
        let result = engine.calculate(
            paneIds: fixture.input.panes.map { $0.id },
            preset: fixture.input.preset,
            containerSize: fixture.input.containerSize
        )
        
        assertLayoutMatches(result: result, expected: fixture.expected)
    }
    
    // MARK: - Preset Tests
    
    func testRowsPreset() {
        let result = engine.calculate(
            paneIds: ["pane-1", "pane-2"],
            preset: .rows,
            containerSize: ContainerSize(width: 800, height: 600)
        )
        
        XCTAssertEqual(result.count, 2)
        
        // First pane on top
        XCTAssertEqual(result[0].id, "pane-1")
        XCTAssertEqual(result[0].x, 0)
        XCTAssertEqual(result[0].y, 0)
        XCTAssertEqual(result[0].width, 800)
        XCTAssertEqual(result[0].height, 300)
        
        // Second pane below
        XCTAssertEqual(result[1].id, "pane-2")
        XCTAssertEqual(result[1].x, 0)
        XCTAssertEqual(result[1].y, 300)
        XCTAssertEqual(result[1].width, 800)
        XCTAssertEqual(result[1].height, 300)
    }
    
    func testMainLeftPreset() {
        let result = engine.calculate(
            paneIds: ["main", "side-1", "side-2"],
            preset: .mainLeft,
            containerSize: ContainerSize(width: 800, height: 600)
        )
        
        XCTAssertEqual(result.count, 3)
        
        // Main pane on left
        XCTAssertEqual(result[0].id, "main")
        XCTAssertEqual(result[0].x, 0)
        XCTAssertEqual(result[0].y, 0)
        XCTAssertEqual(result[0].width, 400)
        XCTAssertEqual(result[0].height, 600)
        
        // Side panes stacked on right
        XCTAssertEqual(result[1].id, "side-1")
        XCTAssertEqual(result[1].x, 400)
        XCTAssertEqual(result[1].y, 0)
        XCTAssertEqual(result[1].width, 400)
        XCTAssertEqual(result[1].height, 300)
        
        XCTAssertEqual(result[2].id, "side-2")
        XCTAssertEqual(result[2].x, 400)
        XCTAssertEqual(result[2].y, 300)
        XCTAssertEqual(result[2].width, 400)
        XCTAssertEqual(result[2].height, 300)
    }
    
    func testMainRightPreset() {
        let result = engine.calculate(
            paneIds: ["main", "side-1", "side-2"],
            preset: .mainRight,
            containerSize: ContainerSize(width: 800, height: 600)
        )
        
        XCTAssertEqual(result.count, 3)
        
        // Side panes stacked on left
        XCTAssertEqual(result[0].id, "side-1")
        XCTAssertEqual(result[0].x, 0)
        XCTAssertEqual(result[0].y, 0)
        XCTAssertEqual(result[0].width, 400)
        XCTAssertEqual(result[0].height, 300)
        
        XCTAssertEqual(result[1].id, "side-2")
        XCTAssertEqual(result[1].x, 0)
        XCTAssertEqual(result[1].y, 300)
        XCTAssertEqual(result[1].width, 400)
        XCTAssertEqual(result[1].height, 300)
        
        // Main pane on right
        XCTAssertEqual(result[2].id, "main")
        XCTAssertEqual(result[2].x, 400)
        XCTAssertEqual(result[2].y, 0)
        XCTAssertEqual(result[2].width, 400)
        XCTAssertEqual(result[2].height, 600)
    }
    
    func testAutoPresetWithOnePane() {
        let result = engine.calculate(
            paneIds: ["pane-1"],
            preset: .auto,
            containerSize: ContainerSize(width: 800, height: 600)
        )
        
        XCTAssertEqual(result.count, 1)
        XCTAssertEqual(result[0].width, 800)
        XCTAssertEqual(result[0].height, 600)
    }
    
    func testAutoPresetWithTwoPanes() {
        let result = engine.calculate(
            paneIds: ["pane-1", "pane-2"],
            preset: .auto,
            containerSize: ContainerSize(width: 800, height: 600)
        )
        
        // Auto with 2 panes = columns
        XCTAssertEqual(result.count, 2)
        XCTAssertEqual(result[0].width, 400)
        XCTAssertEqual(result[1].width, 400)
    }
    
    func testAutoPresetWithThreePanes() {
        let result = engine.calculate(
            paneIds: ["pane-1", "pane-2", "pane-3"],
            preset: .auto,
            containerSize: ContainerSize(width: 800, height: 600)
        )
        
        // Auto with 3 panes = main-left
        XCTAssertEqual(result.count, 3)
        XCTAssertEqual(result[0].width, 400) // main
        XCTAssertEqual(result[1].width, 400) // side
        XCTAssertEqual(result[2].width, 400) // side
    }
    
    func testAutoPresetWithFourPanes() {
        let result = engine.calculate(
            paneIds: ["pane-1", "pane-2", "pane-3", "pane-4"],
            preset: .auto,
            containerSize: ContainerSize(width: 800, height: 600)
        )
        
        // Auto with 4 panes = grid (2x2)
        XCTAssertEqual(result.count, 4)
        for pane in result {
            XCTAssertEqual(pane.width, 400)
            XCTAssertEqual(pane.height, 300)
        }
    }
    
    // MARK: - Zoom Tests
    
    func testZoomPane() {
        let result = engine.calculate(
            paneIds: ["pane-1", "pane-2", "pane-3"],
            preset: .columns,
            containerSize: ContainerSize(width: 800, height: 600),
            zoomedPane: "pane-2"
        )
        
        // When zoomed, only the zoomed pane is returned filling container
        XCTAssertEqual(result.count, 1)
        XCTAssertEqual(result[0].id, "pane-2")
        XCTAssertEqual(result[0].x, 0)
        XCTAssertEqual(result[0].y, 0)
        XCTAssertEqual(result[0].width, 800)
        XCTAssertEqual(result[0].height, 600)
    }
    
    // MARK: - Equalize Tests
    
    func testEqualize() {
        let result = engine.equalize(
            paneIds: ["pane-1", "pane-2"],
            preset: .columns,
            containerSize: ContainerSize(width: 800, height: 600)
        )
        
        // Equalize should reset to default sizing
        XCTAssertEqual(result.count, 2)
        XCTAssertEqual(result[0].width, 400)
        XCTAssertEqual(result[1].width, 400)
    }
    
    // MARK: - Edge Cases
    
    func testEmptyPaneList() {
        let result = engine.calculate(
            paneIds: [],
            preset: .single,
            containerSize: ContainerSize(width: 800, height: 600)
        )
        
        XCTAssertEqual(result.count, 0)
    }
    
    func testGridWithFivePanes() {
        let result = engine.calculate(
            paneIds: ["p1", "p2", "p3", "p4", "p5"],
            preset: .grid,
            containerSize: ContainerSize(width: 900, height: 600)
        )
        
        // 5 panes in grid = 3 cols × 2 rows
        XCTAssertEqual(result.count, 5)
        
        let colWidth = 900.0 / 3.0
        let rowHeight = 600.0 / 2.0
        
        // Verify first row
        XCTAssertEqual(result[0].x, 0)
        XCTAssertEqual(result[0].y, 0)
        XCTAssertEqual(result[0].width, colWidth, accuracy: 0.01)
        XCTAssertEqual(result[0].height, rowHeight, accuracy: 0.01)
        
        XCTAssertEqual(result[1].x, colWidth, accuracy: 0.01)
        XCTAssertEqual(result[1].y, 0)
        
        XCTAssertEqual(result[2].x, colWidth * 2, accuracy: 0.01)
        XCTAssertEqual(result[2].y, 0)
        
        // Verify second row
        XCTAssertEqual(result[3].x, 0)
        XCTAssertEqual(result[3].y, rowHeight, accuracy: 0.01)
        
        XCTAssertEqual(result[4].x, colWidth, accuracy: 0.01)
        XCTAssertEqual(result[4].y, rowHeight, accuracy: 0.01)
    }
    
    // MARK: - Helpers
    
    private func loadLayoutFixture(named name: String) throws -> LayoutFixture {
        // Get the contracts directory (go up from macos/ to project root)
        let testBundle = Bundle(for: type(of: self))
        let projectRoot = testBundle.bundleURL
            .deletingLastPathComponent()  // remove test bundle name
            .deletingLastPathComponent()  // remove .build
            .deletingLastPathComponent()  // remove debug-macos
            .deletingLastPathComponent()  // remove .build
            .deletingLastPathComponent()  // remove macos/
        
        let fixturePath = projectRoot
            .appendingPathComponent("contracts")
            .appendingPathComponent("fixtures")
            .appendingPathComponent("layout")
            .appendingPathComponent("\(name).json")
        
        let data = try Data(contentsOf: fixturePath)
        let decoder = JSONDecoder()
        return try decoder.decode(LayoutFixture.self, from: data)
    }
    
    private func assertLayoutMatches(result: [PaneRect], expected: [ExpectedPaneRect], file: StaticString = #file, line: UInt = #line) {
        XCTAssertEqual(result.count, expected.count, "Pane count mismatch", file: file, line: line)
        
        for expectedPane in expected {
            guard let actualPane = result.first(where: { $0.id == expectedPane.id }) else {
                XCTFail("Missing pane: \(expectedPane.id)", file: file, line: line)
                continue
            }
            
            XCTAssertEqual(actualPane.x, Double(expectedPane.x), accuracy: 0.01, "x mismatch for \(expectedPane.id)", file: file, line: line)
            XCTAssertEqual(actualPane.y, Double(expectedPane.y), accuracy: 0.01, "y mismatch for \(expectedPane.id)", file: file, line: line)
            XCTAssertEqual(actualPane.width, Double(expectedPane.width), accuracy: 0.01, "width mismatch for \(expectedPane.id)", file: file, line: line)
            XCTAssertEqual(actualPane.height, Double(expectedPane.height), accuracy: 0.01, "height mismatch for \(expectedPane.id)", file: file, line: line)
        }
    }
}

// MARK: - Fixture Types

struct LayoutFixture: Codable {
    let name: String
    let input: LayoutInput
    let expected: [ExpectedPaneRect]
}

struct LayoutInput: Codable {
    let containerSize: ContainerSize
    let panes: [PaneId]
    let preset: LayoutPreset
}

struct PaneId: Codable {
    let id: String
}

struct ExpectedPaneRect: Codable {
    let id: String
    let x: Int
    let y: Int
    let width: Int
    let height: Int
}

// Codable conformance for LayoutPreset
extension LayoutPreset: Codable {
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = try container.decode(String.self)
        guard let preset = LayoutPreset(rawValue: rawValue) else {
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "Unknown preset: \(rawValue)")
        }
        self = preset
    }
}

// Codable conformance for ContainerSize
extension ContainerSize: Codable {
    enum CodingKeys: String, CodingKey {
        case width, height
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let width = try container.decode(Double.self, forKey: .width)
        let height = try container.decode(Double.self, forKey: .height)
        self.init(width: width, height: height)
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(width, forKey: .width)
        try container.encode(height, forKey: .height)
    }
}

File: windows/YOLOTerm.Tests/LayoutEngineTests.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using YOLOTerm.Core.Layout;
using Xunit;

namespace YOLOTerm.Tests;

public class LayoutEngineTests
{
    private class LayoutFixture
    {
        [JsonPropertyName("name")]
        public string Name { get; set; } = "";

        [JsonPropertyName("input")]
        public LayoutInput Input { get; set; } = new();

        [JsonPropertyName("expected")]
        public List<ExpectedRect> Expected { get; set; } = new();
    }

    private class LayoutInput
    {
        [JsonPropertyName("containerSize")]
        public ContainerSizeData ContainerSize { get; set; } = new();

        [JsonPropertyName("panes")]
        public List<PaneData> Panes { get; set; } = new();

        [JsonPropertyName("preset")]
        public string Preset { get; set; } = "";

        [JsonPropertyName("zoomedPane")]
        public string? ZoomedPane { get; set; }
    }

    private class ContainerSizeData
    {
        [JsonPropertyName("width")]
        public double Width { get; set; }

        [JsonPropertyName("height")]
        public double Height { get; set; }
    }

    private class PaneData
    {
        [JsonPropertyName("id")]
        public string Id { get; set; } = "";
    }

    private class ExpectedRect
    {
        [JsonPropertyName("id")]
        public string Id { get; set; } = "";

        [JsonPropertyName("x")]
        public double X { get; set; }

        [JsonPropertyName("y")]
        public double Y { get; set; }

        [JsonPropertyName("width")]
        public double Width { get; set; }

        [JsonPropertyName("height")]
        public double Height { get; set; }
    }

    private static string GetContractsPath()
    {
        // Navigate from test assembly to contracts directory
        var testDir = Directory.GetCurrentDirectory();
        var repoRoot = Path.GetFullPath(Path.Combine(testDir, "..", "..", "..", ".."));
        return Path.Combine(repoRoot, "contracts", "fixtures", "layout");
    }

    private static LayoutPreset ParsePreset(string preset)
    {
        return preset.ToLowerInvariant() switch
        {
            "auto" => LayoutPreset.Auto,
            "single" => LayoutPreset.Single,
            "columns" => LayoutPreset.Columns,
            "rows" => LayoutPreset.Rows,
            "grid" => LayoutPreset.Grid,
            "main-left" => LayoutPreset.MainLeft,
            "main-right" => LayoutPreset.MainRight,
            _ => throw new ArgumentException($"Unknown preset: {preset}")
        };
    }

    private static void AssertRectsEqual(List<ExpectedRect> expected, List<PaneRect> actual, string testName)
    {
        Assert.Equal(expected.Count, actual.Count);

        for (int i = 0; i < expected.Count; i++)
        {
            var exp = expected[i];
            var act = actual[i];

            Assert.Equal(exp.Id, act.Id);
            Assert.Equal(exp.X, act.X, precision: 2);
            Assert.Equal(exp.Y, act.Y, precision: 2);
            Assert.Equal(exp.Width, act.Width, precision: 2);
            Assert.Equal(exp.Height, act.Height, precision: 2);
        }
    }

    private void RunFixtureTest(string fixtureFileName)
    {
        var contractsPath = GetContractsPath();
        var fixturePath = Path.Combine(contractsPath, fixtureFileName);

        Assert.True(File.Exists(fixturePath), $"Fixture file not found: {fixturePath}");

        var json = File.ReadAllText(fixturePath);
        var fixture = JsonSerializer.Deserialize<LayoutFixture>(json);
        Assert.NotNull(fixture);

        var engine = new LayoutEngine();
        var paneIds = fixture.Input.Panes.Select(p => p.Id).ToList();
        var containerSize = new ContainerSize(
            fixture.Input.ContainerSize.Width,
            fixture.Input.ContainerSize.Height);
        var preset = ParsePreset(fixture.Input.Preset);

        var result = engine.Calculate(
            paneIds,
            preset,
            containerSize,
            zoomedPane: fixture.Input.ZoomedPane);

        AssertRectsEqual(fixture.Expected, result, fixture.Name);
    }

    [Fact]
    public void SinglePane_FillsContainer()
    {
        RunFixtureTest("single-pane.json");
    }

    [Fact]
    public void TwoColumns_SplitsHorizontally()
    {
        RunFixtureTest("two-columns.json");
    }

    [Fact]
    public void FourGrid_Creates2x2Layout()
    {
        RunFixtureTest("four-grid.json");
    }

    [Fact]
    public void Equalize_ResetsToDefaultSizing()
    {
        var engine = new LayoutEngine();
        var paneIds = new List<string> { "pane-1", "pane-2" };
        var containerSize = new ContainerSize(800, 600);

        var result = engine.Equalize(paneIds, LayoutPreset.Columns, containerSize);

        Assert.Equal(2, result.Count);
        Assert.Equal(0, result[0].X);
        Assert.Equal(400, result[0].Width);
        Assert.Equal(400, result[1].X);
        Assert.Equal(400, result[1].Width);
    }

    [Fact]
    public void ZoomedPane_FillsContainer()
    {
        var engine = new LayoutEngine();
        var paneIds = new List<string> { "pane-1", "pane-2", "pane-3" };
        var containerSize = new ContainerSize(800, 600);

        var result = engine.Calculate(
            paneIds,
            LayoutPreset.Columns,
            containerSize,
            zoomedPane: "pane-2");

        Assert.Single(result);
        Assert.Equal("pane-2", result[0].Id);
        Assert.Equal(0, result[0].X);
        Assert.Equal(0, result[0].Y);
        Assert.Equal(800, result[0].Width);
        Assert.Equal(600, result[0].Height);
    }

    [Fact]
    public void AutoPreset_SinglePane_UsesSingle()
    {
        var engine = new LayoutEngine();
        var paneIds = new List<string> { "pane-1" };
        var containerSize = new ContainerSize(800, 600);

        var result = engine.Calculate(paneIds, LayoutPreset.Auto, containerSize);

        Assert.Single(result);
        Assert.Equal(800, result[0].Width);
        Assert.Equal(600, result[0].Height);
    }

    [Fact]
    public void AutoPreset_TwoPanes_UsesColumns()
    {
        var engine = new LayoutEngine();
        var paneIds = new List<string> { "pane-1", "pane-2" };
        var containerSize = new ContainerSize(800, 600);

        var result = engine.Calculate(paneIds, LayoutPreset.Auto, containerSize);

        Assert.Equal(2, result.Count);
        Assert.Equal(400, result[0].Width);
        Assert.Equal(400, result[1].Width);
    }

    [Fact]
    public void AutoPreset_ThreePanes_UsesMainLeft()
    {
        var engine = new LayoutEngine();
        var paneIds = new List<string> { "pane-1", "pane-2", "pane-3" };
        var containerSize = new ContainerSize(800, 600);

        var result = engine.Calculate(paneIds, LayoutPreset.Auto, containerSize);

        Assert.Equal(3, result.Count);
        // Main pane on left
        Assert.Equal(400, result[0].Width);
        Assert.Equal(600, result[0].Height);
        // Side panes stacked
        Assert.Equal(400, result[1].Width);
        Assert.Equal(300, result[1].Height);
    }

    [Fact]
    public void AutoPreset_FourPanes_UsesGrid()
    {
        var engine = new LayoutEngine();
        var paneIds = new List<string> { "pane-1", "pane-2", "pane-3", "pane-4" };
        var containerSize = new ContainerSize(800, 600);

        var result = engine.Calculate(paneIds, LayoutPreset.Auto, containerSize);

        Assert.Equal(4, result.Count);
        // 2x2 grid
        Assert.All(result, r => Assert.Equal(400, r.Width));
        Assert.All(result, r => Assert.Equal(300, r.Height));
    }

    [Fact]
    public void MainLeft_FirstPaneTakes50Percent()
    {
        var engine = new LayoutEngine();
        var paneIds = new List<string> { "pane-1", "pane-2", "pane-3" };
        var containerSize = new ContainerSize(800, 600);

        var result = engine.Calculate(paneIds, LayoutPreset.MainLeft, containerSize);

        Assert.Equal(3, result.Count);
        Assert.Equal("pane-1", result[0].Id);
        Assert.Equal(400, result[0].Width);
        Assert.Equal(600, result[0].Height);
        Assert.Equal(0, result[0].X);
    }

    [Fact]
    public void MainRight_FirstPaneTakes50PercentOnRight()
    {
        var engine = new LayoutEngine();
        var paneIds = new List<string> { "pane-1", "pane-2", "pane-3" };
        var containerSize = new ContainerSize(800, 600);

        var result = engine.Calculate(paneIds, LayoutPreset.MainRight, containerSize);

        Assert.Equal(3, result.Count);
        // Main pane is last in the list (on the right)
        var mainPane = result.Last();
        Assert.Equal("pane-1", mainPane.Id);
        Assert.Equal(400, mainPane.Width);
        Assert.Equal(600, mainPane.Height);
        Assert.Equal(400, mainPane.X);
    }

    [Fact]
    public void Rows_StacksVertically()
    {
        var engine = new LayoutEngine();
        var paneIds = new List<string> { "pane-1", "pane-2", "pane-3" };
        var containerSize = new ContainerSize(800, 600);

        var result = engine.Calculate(paneIds, LayoutPreset.Rows, containerSize);

        Assert.Equal(3, result.Count);
        Assert.All(result, r => Assert.Equal(800, r.Width));
        Assert.All(result, r => Assert.Equal(200, r.Height));
        Assert.Equal(0, result[0].Y);
        Assert.Equal(200, result[1].Y);
        Assert.Equal(400, result[2].Y);
    }

    [Fact]
    public void EmptyPanes_ReturnsEmptyList()
    {
        var engine = new LayoutEngine();
        var paneIds = new List<string>();
        var containerSize = new ContainerSize(800, 600);

        var result = engine.Calculate(paneIds, LayoutPreset.Columns, containerSize);

        Assert.Empty(result);
    }
}
